### PR TITLE
sys-auth/authenticator: Introduce support for Python 3.7

### PR DESCRIPTION
**Pull request by bucefal91**
Introduce support for Python 3.7 and bump to the latest available version.

This is a raw commit. I've spent hours trying to solve it myself and couldn't... So the problem is: it builds fine but when I run it, it errors as following
```
bucefal91@dell ~ $ authenticator 
Traceback (most recent call last):
  File "/usr/bin/authenticator", line 58, in <module>
    from Authenticator.models import Logger
  File "/usr/lib64/python3.6/site-packages/Authenticator/__init__.py", line 19, in <module>
    from .application import Application
  File "/usr/lib64/python3.6/site-packages/Authenticator/application.py", line 25, in <module>
    from .widgets import Window, AboutDialog
  File "/usr/lib64/python3.6/site-packages/Authenticator/widgets/__init__.py", line 20, in <module>
    from .accounts import AccountsList, AccountRow, AccountsListState
  File "/usr/lib64/python3.6/site-packages/Authenticator/widgets/accounts/__init__.py", line 19, in <module>
    from .add import AddAccountWindow
  File "/usr/lib64/python3.6/site-packages/Authenticator/widgets/accounts/add.py", line 28, in <module>
    from ..headerbar import HeaderBarButton
  File "/usr/lib64/python3.6/site-packages/Authenticator/widgets/headerbar.py", line 27, in <module>
    from ..models import Database
  File "/usr/lib64/python3.6/site-packages/Authenticator/models/__init__.py", line 26, in <module>
    from .qr_reader import QRReader
  File "/usr/lib64/python3.6/site-packages/Authenticator/models/qr_reader.py", line 21, in <module>
    from PIL import Image
  File "/usr/lib64/python3.6/site-packages/PIL/Image.py", line 64, in <module>
    from . import _imaging as core
ImportError: cannot import name '_imaging'
```

Funny enough, when I try execute that import statement from python CLI, it works!
```
bucefal91@dell ~ $ python3.6
Python 3.6.6 (default, Oct 16 2018, 15:28:12) 
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from PIL import Image
>>> 
```

I've researched online for this kind of behavior but did not find any answer. Unfortunately I am not a python developer, so my capacity is limited.

Closes: https://bugs.gentoo.org/663162
Package-Manager: Portage-2.3.51, Repoman-2.3.12
Signed-off-by: Oleksandr Trotsenko <oleksandr.trotsenko@gmail.com>

### CODE REVIEW

**Comment by juippis** [comment]
So, you're trying to get an ebuild to the tree that is not working? :) 
There are many places for support and you should probably ask there first. For example in the bug, you can join IRC and #gentoo-dev-help or #gentoo-python, forums...

But since you've already opened this I guess it's a bit late. First, I see tests are enabled, did you run them? They could tell you something. 

Then: you might need to fix python shebang in /usr/bin/authenticator script (https://github.com/bilelmoussaoui/Authenticator/blob/master/authenticator.py.in seems to just call python3, which uses system default -> in your case, the system default is 3.6 and you installed this using 3.7) or make sure user's system default is the same as chosen python single impl  in here.

Then: I see huge changes in the ebuild between 2.1 and 2.4, don't you want to update -9999 too? Does it still work? :)

You might need to amend the commit and force-push to get the CI bot to rerun this commit again. Although you'll need to fix this contribution, so you gonna have to rebase, squash and force-push anyway.

And don't worry, lot of text from me, but this **should** be quite easily fixable.

File: sys-auth/authenticator/authenticator-0.2.4.ebuild [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+RDEPEND="
+	${PYTHON_DEPS}
+	app-crypt/libsecret
+	dev-python/pillow[${PYTHON_USEDEP}]
+	dev-python/pygobject:3[${PYTHON_USEDEP}]
+	dev-python/pyotp[${PYTHON_USEDEP}]
+	dev-python/python-gnupg[${PYTHON_USEDEP}]
+	dev-python/pyzbar[${PYTHON_USEDEP}]
+	x11-libs/gtk+:3
+	qrcode? (
```
Comment (juippis) [inline sys-auth/authenticator/authenticator-0.2.4.ebuild]: (Opinion, but to me, this is more readable as a one-liner:
`qrcode? ( media-gfx/gnome-screenshot )`

but it's a personal style thing and you can do it how you want, I'm just typing this if you weren't aware it can go to one line)

File: sys-auth/authenticator/authenticator-0.2.4.ebuild [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+		media-gfx/gnome-screenshot
+	)
+"
+DEPEND="${RDEPEND}"
+
+pkg_preinst() {
```
Comment (juippis) [inline sys-auth/authenticator/authenticator-0.2.4.ebuild]: You should re-arrange these functions to match the order they get executed. Granted this works, but it's harder for others to read.
https://devmanual.gentoo.org/ebuild-writing/functions/index.html

File: sys-auth/authenticator/authenticator-0.2.4.ebuild [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+pkg_postrm() {
+	gnome2_icon_cache_update
+	gnome2_schemas_update
+}
+
+src_prepare() {
+	rm -r "subprojects/libgd" || die "Could not remove the subproject libgd folder."
+	ln -s "../../libgd-${GNOME_LIBGD_COMMIT}" "subprojects/libgd" || die "Could not symlink libgd subproject."
+
+	default_src_prepare
```
Comment (juippis) [inline sys-auth/authenticator/authenticator-0.2.4.ebuild]: You can just shorten this to `default`.

File: sys-auth/authenticator/metadata.xml [hunk 4]
```diff
@@ -9,6 +9,9 @@
 		<email>proxy-maint@gentoo.org</email>
 		<name>Proxy Maintainers</name>
 	</maintainer>
+	<use>
+		<flag name="qrcode">Enable scanning of QR codes.</flag>
```
Comment (juippis) [inline sys-auth/authenticator/metadata.xml]: `.` is not needed for consistency in the ebuild tree.

File: sys-auth/authenticator/authenticator-0.2.4.ebuild (hunk 3 above)
Comment (juippis) [reply sys-auth/authenticator/authenticator-0.2.4.ebuild]: Also it should be the first thing to call in src_prepare, before rm and ln

**Comment by bucefal91** [comment]
@juippis Thank you :) Yes, I was going to request help on IRC with this hiccup. First I wanted to push the code out and made a PR since the code was pushed.

I've gone through your feedback, just force pushed an amend to my commit. I appreciate it :) The `-9999` ebuild is yet to come... upstream introduced one extra dependency since 0.2.4 version so I am trying to keep it scoped.. First to roll out 0.2.4 and then see about that extra depedency and thus handle -9999.ebuild.

I am heading now to gentoo's IRC channels for help :)

**Comment by candrews** [comment]
@bucefal91 as proxy maintainer, can you please take a look at this?

**Comment by candrews** [comment]
@bucefal91 does this work now? Did you solve the problem with it not running? And do all tests pass?

**Comment by bucefal91** [comment]
So here's more updates.. Just pushed an amend to my commit in this pull request.

After checking out QR codes, turns out I just had a wrong QR code on which I was running my tests last time. Now I confirm QR scanner works within Authenticator

Also, on paper they do define explicit additional dependencies for QR scanner https://gitlab.gnome.org/World/Authenticator but at least in 0.2.4 version the application wouldn't start unless the QR code dependencies are also installed. So I opted for removing the USE flag since it was appearing pointless.

Tests pass locally (though they are very limited). and I confirm the app works in general as I am using it now for my 2fa on bitbucket.org for more than 1 week.

There was also a migration from github to GNOME's facilities (https://gitlab.gnome.org). However, I do confirm MANIFEST checksum for 0.2.4 version from github and gitlab.gnome.org match. Thus I deem gitlab.gnome.org trustworthy source of code. Authenticator is being developed now in the gnome repository, so I updated the project's homepage (and a few other spots) to reflect the migration from github onto GNOME's gitlab.

All in all, now I deem this ebuild worth reviewing by others and ready for merge unless something comes up during the review.

**Comment by candrews** [comment]
I'm not comfortable with the bundled libgd; please depend on media-libs/gd instead.

**Comment by candrews** [comment]
pushed with fixes https://github.com/gentoo/gentoo/commit/a0508b6bad045c67be5c8a0ce7fed393bb199c43

**Comment by bucefal91** [comment]
@candrews , thank you for reviewing, feedback, fixing, and committing!
To be honest by now I do not recall why I couldn't unbundle the gd library myself. Though I do remember I invested some few hours into it and eventually gave up.

But anyway, I am very pleased to see this committed!

**Comment by candrews** [comment]
I'm glad to help - please feel free to ping me in the future if you need help, have questions, or just want something moved along.
